include/ofi_util.h

```diff
@@ -162,6 +162,21 @@ struct util_prov {
 	const struct fi_info		*info;
 	ofi_alter_info_t		alter_defaults;
 	const int			flags;
+
+	/* Bit mask of optional capabilities which can be returned in the caps
+	 * fields. Recommendation is to set this to caps which providers can
+	 * optimally support. Setting to zero will result in no optional
+	 * capabilities being returned in caps.
+	 */
+	uint64_t			opt_caps_valid_caps;
+
+	/* Bit mask of optional capabilities which can be returned in
+	 * optional_cap fields. Recommendation is to set this to caps which
+	 * providers have to emulate and/or cannot support optimally. Setting to
+	 * zero will result in no optional capabilities being returned in
+	 * optional caps.
+	 */
+	uint64_t			opt_caps_valid_opt_caps;
```

Comment:
Providers can specify optional_caps in their base fi_info's.  I think we should make use of those structs, rather than assuming the optional caps apply provider wide.  They may apply per fi_info.

We could have the provider set optional_caps be either the recommended or not recommended caps and update the util code accordingly.  (I would go with not recommended).  The info->caps would continue to be the full set of caps, same as today.

This path likely reduces the changes needed to the util code.  If optional_caps = 0, the current behavior is used, so we don't have to update all providers.  If optional_caps is set, then the util code would adjust the returned info->caps based on hints->caps and hints->optional_caps, removing those flags set in the provider info->optional_caps and not in hints->caps.

Reply:
I like the idea of using providers fi_info::optional_caps field. The issue is I need to know when optional_caps are supported and which optional_caps should be returned through fi_info::caps or fi_info::optional_caps. Thoughts on the below behavior?

```
/* Enable/disable optional caps. This would be global per provider.
bool optional_caps_enabled;

/* Per provider fi_info::optional_caps value. If optional_caps are supported, this
 * defines the optional capabilities which can only be returned in fi_info::optional_caps.
 * All other optional capabilities are eligible to be returned in fi_info::caps.
 */
bool prov_optional_caps;

/* All the capabilities supported by the provider. */
uint64_t prov_caps;

/* User input from fi_info hints. */
uint64_t hints_optional_caps;
uint64_t hints_caps;

/* fi_info::caps and fi_info::optional_caps returned to users. */
uint64_t info_caps = 0;
uint64_t info_optional_caps = 0;

if (optional_caps_enabled) {
	uint64_t requested_caps = hints_caps | hints_optional_caps;
	uint64_t supported_caps = requested_caps & prov_caps;
	
	info_caps = supported_caps & ~prov_optional_caps;
	info_optional_caps = supported_caps & prov_optional_caps;
} else {
	info_caps = hints_caps;
}

return info_caps, info_optional_caps
```